libqtile/group.py

```diff
@@ -83,23 +85,54 @@ def current_window(self):
 
     @current_window.setter
     def current_window(self, win):
-        try:
+        old = self._order_clients()
+        if win in self.focus_history:
             self.focus_history.remove(win)
-        except ValueError:
-            # win has never received focus before
-            pass
+        self.windows.remove(win)
+
         self.focus_history.append(win)
+        self.windows.append(win)
+        if old != self._order_clients():
+            self.assure_correct_layer(win)
 
-    def _remove_from_focus_history(self, win):
-        try:
-            index = self.focus_history.index(win)
-        except ValueError:
-            # win has never received focus
-            return False
+    def _order_clients(self):
+        """
+        Orders our clients by their z-value, leaving the order as is in case of equal z.
+        """
+        def sort_by(client):
+            if client.floating and not client.fullscreen:
+                return client.z + window.FLOAT_LAYER
+            return client.z
+        return sorted(filter(lambda w: not w.minimized, self.windows), key=sort_by)
+
+    def assure_correct_layer(self, win, **kwargs):
+        """
+        Make sure that the window is in its designated z-position by moving other windows around as necessary.
+        We assume that win is already at the correct position in our list of windows. Nobody else should be
+        calling xcbq.configure with a StackMode (except Popups).
+        """
+        if win.minimized:
+            return
+
+        stacked = ['max']  # z-order in these layouts does matter
+        windows = self._order_clients()
+
+        # try to avoid moving other windows at all
+        if not win.floating and not win.fullscreen and self.layout.name not in stacked:
+            win.window.configure(stackmode=StackMode.Below, **kwargs)
+            return
+
+        index = windows.index(win)
+        # decide in which direction we have less windows to draw
+        if True:  # index >= len(windows) / 2:
```

Comment:
I was hoping to find a way to disallow focus changes while this code is being run, since sometimes the `else`-path will be faster. I haven't, so I'll remove the other path.